docs/design/object-locking.md

```diff
@@ -0,0 +1,187 @@
+`retain-until-date` date and time to be stored on the object's metadata.
+Lock configuration on an object version can also be changed using
+`PutObjectRetention` api request, when the object has no prior retention mode
+set or if the object has `GOVERNANCE` mode set and the user has appropriate
+permission to make the request.
+
+Note: Delete markers do not have any object lock protection
+
+#### Processing DELETE requests
+
+Whenever requests such as  DELETE object using version-id,
+Multi-Object DELETE specifying version-id or a Lifecycle action to
+permanently delete the object version are received, the current date and time is
+evaluated against `retain-until-date` set on the object and the client will
+receive `Access Denied` error if the current date and time is less than the
+`retain-until-date` set on the object.
+When the current date and time exceeds the `retain-until-date`, deletes on the
+object version are automatically allowed. There is no cleanup action to remove
+the `retain-until-date` set on the object version's metadata once the
+`retain-until-date` expires.
+DELETE object requests without a version id result in creation of delete markers
+on top of the object version, even if the object has a lock configuration is set
+and the `retain-until-date` is current.
+
+#### Lifecycle actions
+
+Lifecycle jobs can create delete markers on the object but cannot delete an
+object version permanently that is locked until the lock expires.
+
+#### Replication
+
+Lock configuration on the object version in a source bucket is copied over
+to the destination bucket only if object lock is enabled on the destination
+bucket, otherwise the lock is ignored in the destination bucket.
+
+## APIs covering S3 Object lock
+
+* **Put Bucket** - extend bucket creation api to include configuration for
+  enabling object lock on the bucket
+
+  **Note:** Versioning is automatically enabled on buckets that have object lock
+  enabled as part of PUT Bucket request
+
+* **PUT Object** - extend put object api to parse `x-amz-object-lock-mode`,
+  `x-amz-object-lock-retain-until-date`, `x-amz-object-lock-legal-hold`
+  headers and store the configuration on an object.
+
+* **COPY Object** - extend the api to accept the similar lock configuration
+  headers as the PUT Object request
+
+* **Create Multipart Upload** - extend the api to accept the similar lock
+  configuration headers as the PUT Object request
+
+* **Put Object Lock Configuration** - allows setting a default lock
+  configuration for objects that are going to be stored in the bucket. This
+  request is accepted only on buckets that have object lock enabled.
+
+* **Get Object Lock Configuration** - GET the object lock configuration set on
+  the bucket metadata
+
+* **Put Object Retention** - sets retention mode/period configuration on a
+  object version
+
+* **Get Object Retention**- get retention mode/period configuration set on a
+  object version
+
```

Comment:
Are we supposed to maintain UTAPI metrics ?
Are we supposed to maintain statistics counters for monitoring attempts to delete a lock (security monitoring) ?

Reply:
We will include hooks to Utapi but beyond that I don't think Utapi bodes well for audit trail. We have another feature competing in S3C for Audit trail and I think it's best to introduce it there.